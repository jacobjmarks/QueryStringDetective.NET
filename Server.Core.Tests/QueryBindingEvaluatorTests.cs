namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "c",
        "foo",
        "a&q=b&q=c",
        "null",
        "0",
        "1",
        "-1",
        "true",
        "false",
        "yes",
        "no",
        "1&q=2&q=3",
        "1&q=2&q=.3",
        "-1&q=0&q=1",
        $"{sbyte.MinValue}",
        $"{sbyte.MaxValue}",
        $"{byte.MaxValue}",
        $"{short.MinValue}",
        $"{short.MaxValue}",
        $"{ushort.MaxValue}",
        $"{int.MinValue}",
        $"{int.MaxValue}",
        $"{long.MinValue}",
        $"{long.MaxValue}",
        $"{ulong.MaxValue}",
        Uri.EscapeDataString($"{double.MinValue}"),
        Uri.EscapeDataString($"{double.MaxValue}"),
        $"{decimal.MinValue}",
        $"{decimal.MaxValue}",
        Uri.EscapeDataString($"{float.MinValue}"),
        Uri.EscapeDataString($"{float.MaxValue}"),
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?q=" + queryString);

        var content = $"Input: {queryString}"
            + "\n\nBindings:\n";

        foreach (var result in results)
        {
            content += "\n" + result.Type;
            content += "\n  " + string.Join("\n  ", result.Results
                .Select(r => $"{r.Key}\n    {(
                    r.Value.IsErroneous
                    ? $"e = {r.Value.Error.Message} ({r.Value.Error.Detail})"
                    : $"r = {r.Value.Result}")}"));

            content += "\n";
        }

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}