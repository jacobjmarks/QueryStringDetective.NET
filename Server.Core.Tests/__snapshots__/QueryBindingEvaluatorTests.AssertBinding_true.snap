Input: true

Bindings:

string
  MinimalApis
    r = "true"

string[]
  MinimalApis
    r = ["true"]

char
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char q" from "true".)

char[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char[] q" from "true".)

int
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int q" from "true".)

int[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int[] q" from "true".)

double
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "double q" from "true".)

double[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "double[] q" from "true".)

float
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "float q" from "true".)

float[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "float[] q" from "true".)

decimal
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "decimal q" from "true".)

decimal[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "decimal[] q" from "true".)

bool
  MinimalApis
    r = true

bool[]
  MinimalApis
    r = [true]

uint
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint q" from "true".)

uint[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint[] q" from "true".)

long
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long q" from "true".)

long[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long[] q" from "true".)

ulong
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong q" from "true".)

ulong[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong[] q" from "true".)

sbyte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte q" from "true".)

sbyte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte[] q" from "true".)

short
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short q" from "true".)

short[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short[] q" from "true".)

ushort
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort q" from "true".)

ushort[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort[] q" from "true".)

byte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte q" from "true".)

byte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte[] q" from "true".)
